x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/expression/function/scalar/string/TopSnippets.java

```diff
@@ -0,0 +1,241 @@
+        detailedDescription = """
+                TopSnippets can be used on fields from the text famiy like <<text, text>> and <<semantic-text, semantic_text>>.
+                TopSnippets will extract the best snippets for a given query string.
+            """,
+        examples = {
+            @Example(file = "top-snippets", tag = "top-snippets", applies_to = "stack: preview 9.3.0"),
+            @Example(file = "top-snippets", tag = "top-snippets-with-options", applies_to = "stack: preview 9.3.0") }
+    )
+    public TopSnippets(
+        Source source,
+        @Param(name = "field", type = { "keyword", "text" }, description = "The input to chunk.") Expression field,
+        @Param(name = "query", type = { "keyword" }, description = "The query string to extract the best snippets for.") Expression query,
+        @MapParam(
+            name = "options",
+            description = "Options to customize snippet extraction behavior.",
+            optional = true,
+            params = {
+                @MapParam.MapParamEntry(
+                    name = "num_snippets",
+                    type = { "integer" },
+                    description = "The maximum number of matching snippets to return.",
+                    valueHint = { "3" }
+                ),
+                @MapParam.MapParamEntry(name = "num_words", type = { "integer" }, description = """
+                    The maximum number of words to return in each snippet.
+                    This allows better control of inference costs by limiting the size of tokens per snippet.
+                    """, valueHint = { "300" }) }
+        ) Expression options
+    ) {
+        super(source, options == null ? List.of(field, query) : List.of(field, query, options));
+        this.field = field;
+        this.query = query;
+        this.options = options;
+    }
+
+    public TopSnippets(StreamInput in) throws IOException {
+        this(
+            Source.readFrom((PlanStreamInput) in),
+            in.readNamedWriteable(Expression.class),
+            in.readNamedWriteable(Expression.class),
+            in.readOptionalNamedWriteable(Expression.class)
+        );
+    }
+
+    @Override
+    public void writeTo(StreamOutput out) throws IOException {
+        source().writeTo(out);
+        out.writeNamedWriteable(field);
+        out.writeNamedWriteable(query);
+        out.writeOptionalNamedWriteable(options);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return ENTRY.name;
+    }
+
+    @Override
+    public DataType dataType() {
+        return DataType.KEYWORD;
+    }
+
+    @Override
+    protected TypeResolution resolveType() {
+        if (childrenResolved() == false) {
+            return new TypeResolution("Unresolved children");
+        }
+
+        return isString(field(), sourceText(), FIRST).and(() -> isString(query(), sourceText(), SECOND))
```

Comment:
I think we want the field to be NULL as well as a String.

If a query is done over a field that doesn't exist, it will be replaced by NULL - so I think snippets over NULL should just return null, but allow it as a type.